### PR TITLE
Add support for --acknowledge-abuse flag

### LINKED ISSUE #399: --drive-acknowledge-abuse Equivalent in Plexdrive?
is there a command line switch that allows files marked as spam or malware by google drive to still be downloaded in Plexdrive?  something similar to the RClone --drive-acknowledge-abuse switch?

the current command line is plexdrive mount -v 3 --chunk-check-threads=16 --chunk-load-threads=16 --chunk-load-ahead=16 --max-chunks=256 /home/user/mountedfolder/

and some files get 403 errors "cannotdownloadabusivefile"

thanks

### PR DESCRIPTION
**Pull request by raidancampbell**
This resolves #399 via an additional `--acknowledge-abuse` flag.  Right off the bat, I totally understand if this isn't a desired feature: the comment in #399 was ambiguous if it's desired.  If you don't want it, feel free to close.

The implementation itself is straightforward and comes from [the documentation](https://developers.google.com/drive/api/v3/manage-downloads).  I've briefly tested it and confirm the flag is sent, but I don't have any setup affected by the bug so can't confirm the fix.

### CODE REVIEW

File: chunk/download.go [hunk 1]
```diff
@@ -67,6 +69,9 @@ func (d *Downloader) thread() {
 
 func (d *Downloader) download(client *http.Client, req *Request, buffer []byte) {
 	Log.Debugf("Starting download %v (preload: %v)", req.id, req.preload)
+	if d.acknowledgeAbuse {
+		req.acknowledgeAbuse = true
```
Comment (felixbuenemann) [inline chunk/download.go:73]: You shouldn't modify the Request struct passed as input here, instead set `acknowledgeAbuse` when the struct ist built in `requestChunk()` of the manager.

File: chunk/download.go [hunk 2]
```diff
@@ -97,7 +102,11 @@ func downloadFromAPI(client *http.Client, request *Request, buffer []byte, delay
 		Log.Debugf("%v", err)
 		return fmt.Errorf("Could not create request object %v (%v) from API", request.object.ObjectID, request.object.Name)
 	}
-
+	if request.acknowledgeAbuse {
+		q := req.URL.Query()
+		q.Add("acknowledgeAbuse", "true")
+		req.URL.RawQuery = q.Encode()
+	}
```
Comment (felixbuenemann) [inline chunk/download.go:109]: Instead of decoding the query of the request, adding a query param and then reencoding it, why not store a copy of `DownloadURL` and append `"&acknowledgeAbuse=true"` to it, before building the http request?

We know that all download urls already have at least "alt=media" query param, so no need to handle different cases.

**Review by felixbuenemann — CHANGES_REQUESTED** [review]

File: chunk/download.go (hunk 1 above)
Comment (raidancampbell) [reply chunk/download.go]: resolved via 441ad17d0defde090917b837265f6ce06aea80e9

File: chunk/download.go (hunk 2 above)
Comment (raidancampbell) [reply chunk/download.go]: It felt less hacky than appending a fixed string to the URL, but it's not a strong opinion I have. 
 12a93020718bddec7427647e9d69636b101a2e08 changes this to concatenate the fixed string query parameter to the URL

File: chunk/download.go [hunk 3]
```diff
@@ -92,12 +92,14 @@ func downloadFromAPI(client *http.Client, request *Request, buffer []byte, delay
 		time.Sleep(time.Duration(delay) * time.Second)
 	}
 
+	if request.acknowledgeAbuse {
+		request.object.DownloadURL += "&acknowledgeAbuse=true"
+	}
```
Comment (felixbuenemann) [inline chunk/download.go:97]: Please work with a copy of `DownloadURL`, you are modifying the request struct, that was passed into the function as a parameter.

```go
downloadURL := request.object.DownloadURL
if request.acknowledgeAbuse {
        downloadURL += "&acknowledgeAbuse=true"
}
req, err := http.NewRequest("GET", downloadURL, nil)
```

File: chunk/manager.go [hunk 4]
```diff
@@ -13,6 +13,7 @@ type Manager struct {
 	downloader *Downloader
 	storage    *Storage
 	queue      chan *QueueEntry
+	acknowledgeAbuse bool
```
Comment (felixbuenemann) [inline chunk/manager.go:16]: Please fix indentation of the struct definition.

File: chunk/manager.go [hunk 5]
```diff
@@ -71,6 +67,7 @@ func NewManager(
 		downloader: downloader,
 		storage:    storage,
 		queue:      make(chan *QueueEntry, 100),
+		acknowledgeAbuse: ackAbuse,
```
Comment (felixbuenemann) [inline chunk/manager.go:70]: Please fix indentation with other parameters.

File: chunk/manager.go [hunk 6]
```diff
@@ -136,6 +133,7 @@ func (m *Manager) requestChunk(object *drive.APIObject, offset, size int64, sequ
 		chunkOffsetEnd: chunkOffset + size,
 		sequence:       sequence,
 		preload:        false,
+		acknowledgeAbuse: m.acknowledgeAbuse,
```
Comment (felixbuenemann) [inline chunk/manager.go:136]: Please fix indentation with other parameters.

File: chunk/download.go [hunk 7]
```diff
@@ -15,24 +15,24 @@ import (
...
-	queue      chan *Request
-	callbacks  map[string][]DownloadCallback
-	lock       sync.Mutex
-	storage    *Storage
+	Client           *drive.Client
+	BufferSize       int64
+	queue            chan *Request
+	callbacks        map[string][]DownloadCallback
+	lock             sync.Mutex
+	storage          *Storage
```
Comment (felixbuenemann) [inline chunk/download.go:23]: Please fix indentation.

File: chunk/download.go [hunk 8]
```diff
@@ -15,24 +15,24 @@ import (
...
-		Client:     client,
-		BufferSize: bufferSize,
-		queue:      make(chan *Request, 100),
-		callbacks:  make(map[string][]DownloadCallback, 100),
-		storage:    storage,
+		Client:           client,
+		BufferSize:       bufferSize,
+		queue:            make(chan *Request, 100),
+		callbacks:        make(map[string][]DownloadCallback, 100),
+		storage:          storage,
```
Comment (felixbuenemann) [inline chunk/download.go:35]: Please fix indentation.

**Review by felixbuenemann — CHANGES_REQUESTED** [review]

File: chunk/download.go (hunk 3 above)
Comment (raidancampbell) [reply chunk/download.go:98]: Apologies for misunderstanding your initial request.  Implemented in 287938feea0b1246bea1977dc59e9f2ef003c4d6

File: chunk/manager.go (hunk 4 above)
Comment (raidancampbell) [reply chunk/manager.go:16]: well that's embarrassing.  287938feea0b1246bea1977dc59e9f2ef003c4d6 runs `gofmt` on `manager.go` and `download.go`

**Review by felixbuenemann — APPROVED** [review]

**Comment by felixbuenemann** [comment]
@raidancampbell Thank you for working with me on this. Merged to master.

**Comment by Moodkiller** [comment]
Found a small bug with this on the latest version (5.2.1), if the file is NOT flagged but the `--acknowledge-abuse` is set, plexdrive won't download the file: 

image

**Comment by felixbuenemann** [comment]
The current logic is wrong. We need to try the download without acknowledgeAbuse first, and retry with the flag if we get a 403 with error cannotDownloadAbusiveFile.

**Comment by felixbuenemann** [comment]
@Moodkiller @raidancampbell I've tried to fix this in https://github.com/felixbuenemann/plexdrive/tree/fix-acknowledge-abuse-download-of-unflagged-files. The fix isn't perfect, since it means we do twice as many download requests for abused files, but remembering this for each file would require much more code changes.

Please try how well this works for you for both file flagged for abuse and not flagged by building from the branch linked above or using the attached plexdrive binary (linux-amd64 only): [plexdrive-ackabuse-fix-linux-amd64.zip](https://github.com/plexdrive/plexdrive/files/8010450/plexdrive-ackabuse-fix-linux-amd64.zip)

If it works fine for you, I'll do a new bugfix release.

**Comment by Moodkiller** [comment]
Well, I can at least confirm that running with `--acknowledge-abuse` allows **non** flagged files to be downloaded. ~~I don't have any flagged at this stage to test but will be sure to do so if I come across one.~~

flags are: `-o allow_other -v 3 -o direct_io --chunk-check-threads=2 --chunk-load-ahead=3 --chunk-load-threads=2 --chunk-size=10M --max-chunks=64 --refresh-interval=5m --acknowledge-abuse`

Thank you for the fast fix!

EDIT: updated comment here: https://github.com/plexdrive/plexdrive/pull/410#pullrequestreview-874156752.
TL:DR, flagged files still dont download even with plexdrive flag set.
Would it also be possible to add a logic that eventually skips / fails the download all together?
